In [1]:
import pandas as pd
import numpy as np
import sqlite3

In [2]:
df = pd.read_csv('dataset/train.csv')
df
#loading in the csv file into my system, seeing the initial length before dropping initial values

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


In [3]:
column_names = df.columns.tolist()
column_names
#getting a list of column names so I know which one to get rid of

['PassengerId',
 'Survived',
 'Pclass',
 'Name',
 'Sex',
 'Age',
 'SibSp',
 'Parch',
 'Ticket',
 'Fare',
 'Cabin',
 'Embarked']

In [4]:
del df['Cabin']
#note: I was hesitant to get rid of Cabin, but decided to delete it because nearly 80% of it was empty values

In [5]:
df
#making sure that the values are correct and Cabin is gone

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,S
...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C


In [6]:
median_value = df['Age'].median()
#to fill in the empty values, I chose to fill in all the 'NaN' values with the median 
df['Age'] = df['Age'].fillna(median_value)
df

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,S
...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,28.0,1,2,W./C. 6607,23.4500,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C


In [7]:
df['Embarked'] = df['Embarked'].fillna("S")
#making sure to filling any remaining values

In [8]:
df['Sex'] = df['Sex'].astype('category')

In [9]:
names = df['Name']

df['Title'] = df['Name'].str.extract(r'([A-Za-z]+)\.')
df
#creating a new category for the title, which will help us for finally making our analysis

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,Title
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,S,Mr
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C,Mrs
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S,Miss
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S,Mrs
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,S,Mr
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,S,Rev
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,S,Miss
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,28.0,1,2,W./C. 6607,23.4500,S,Miss
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C,Mr


In [11]:
conn = sqlite3.connect("titanic.db")
result = pd.read_sql_query("SELECT COUNT(*) FROM passengers", conn)
result

,COUNT(*)
0,891


In [12]:
result = pd.read_sql_query("""
    SELECT Pclass, AVG(Survived) as survival_rate
    FROM passengers
    GROUP BY Pclass
""", conn)
result

,Pclass,survival_rate
0,1,0.629630
1,2,0.472826
2,3,0.242363


In [13]:
result = pd.read_sql_query("""
    SELECT Title, AVG(Survived) as survival_rate, COUNT(*) as count
    FROM passengers
    GROUP BY Title
    ORDER BY count DESC
""", conn)
result

,Title,survival_rate,count
0,Mr,0.156673,517
1,Miss,0.697802,182
2,Mrs,0.792000,125
3,Master,0.575000,40
4,Dr,0.428571,7
5,Rev,0.000000,6
6,Mlle,1.000000,2
7,Major,0.500000,2
8,Col,0.500000,2
9,Sir,1.000000,1


This ETL pipeline cleans the Titanic passenger dataset and loads it into SQLite for analysis. I dropped the Cabin column after running early tests because 77% of its values were missing, making it unreliable and impossible to fill in a meaningful way. I replaced missing Age values with the median age, since I felt a value in line with the existing trend in the data would introduce less distortion than the mean, which is more sensitive to outliers. I extracted a Title column from the Name field (Mr/Mrs/Miss/Master/etc.) because titles capture demographic and social information that Sex and Age alone don't fully represent (for example, distinguishing young boys ("Master") from adult men ("Mr"), or unmarried women ("Miss") from married women ("Mrs")).

My three verification queries show: 
(1) a row count of 891, confirming no data was lost or duplicated during the load step; 
(2) survival rate by Pclass, which shows a clear downward trend as class decreases (1st class ~63%, 2nd class ~47%, 3rd class ~24%)
(3) survival rate by Title, which shows that despite adult men ("Mr") making up the largest group of passengers (517 of 891), they had by far the lowest survival rate (~16%)

While women and young boys survived at much higher rates, reflecting the "women and children first" evacuation pattern by outcome, they did not make up the passenger composition.